lib/bitvec.cpp

```diff
@@ -145,3 +146,20 @@ bool bitvec::is_contiguous() const {
         return false;
     return max().index() - min().index() + 1 == popcount();
 }
+
+/**
+ * The purpose of this function is to barrel_shift a bitvec around a pivot.  This function
+ * will crash if the pivot is smaller than the size of the bitvector
+ */
+bitvec bitvec::barrel_shift_left(size_t shift, size_t pivot) const {
+    if (max().index() > pivot) {
+        BUG("Barrel shifting around pivot %d is smaller than the max of the bitvec %d", pivot,
+            max().index());
+        return *this;
+    }
+    int single_shift = shift % pivot;
+    bitvec rv;
+    rv |= *this << single_shift;
+    rv |= *this >> (pivot - single_shift);
+    return rv & bitvec(0, pivot);
```

Comment:
`rv.clrrange(pivot, ~0);` would be slightly more efficient for this last step (in place clear rather than creating a bitmask and anding)